Bug fix of no space in the result of to_string.

Telomelonia opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

As mentioned in the issue, in the result of the to_string() function, there is no space between value and unit.So, I added a space between value and unit.

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address ...

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #7455 

Close https://github.com/astropy/astropy/pull/14231

TODOS
- [x] space in LaTex format 
- [x] fix for format = generic and unicode cases.
- [x] centralize the bug fix on the unit rather than the format.

pllim commented on docs/changes/coordinates/14379.bugfix.rst:
```diff
@@ -0,0 +1 @@
+The bug is fixed in ``to_string``.
```
You would have to elaborate what "the bug" is here. This will be part of the change log for the whole package.

mhvk commented on docs/changes/coordinates/14379.bugfix.rst:
```diff
@@ -0,0 +1,2 @@
+In result of ``to_string``, there is no space between value and unit.
```
Sorry, realize my suggestion for the changelog message was actually wrong -- now edited above to give something that's hopefully correct!

eerovaher reviewed (changes requested):
With the changes this pull request is proposing the following works as it should:
```python
>>> from astropy.coordinates import Angle
>>> Angle(5, unit="rad").to_string()
'5 rad'
```
But the following still produce a string without a space (it would be good to add a unit test analogous to one of these):
```python
>>> Angle(5, unit="rad").to_string(format="unicode")
'5rad'
>>> Angle(5, unit="rad").to_string(format="generic")
'5rad'
```
It would seem that a space should not be added between the value and the unit if the format is `"latex"` or `"latex_inline"`, or if the output is sexagesimal.

mhvk commented:
Oops, indeed, this is more of a partial fix than I had realized! I think in principle there should be a space also for `latex`, etc. In #14231, I had [suggested](https://github.com/astropy/astropy/pull/14231/files#r1057999266) to use `super()` for the non-sexagesimal case, but a quick try caused more breakage than I had thought would be the case. Also `Quantity.to_string()` does not support `unicode` (see new issue #14386) and arguably does `latex` wrong in typesetting the degree symbol with a space in between it and the value (new issue #14385).

One option here may be to just keep it simple and focus on the generic format only -- i.e., only update this PR to also allow `format='generic'`.

Telomelonia commented on astropy/coordinates/tests/test_angles.py:
```diff
@@ -1096,8 +1096,8 @@ def parse_test(i=0):
         (np.nan * u.rad, "nan", "nan rad"),
-        ([np.nan, 1, 0] * u.rad, "[nan 1rad 0rad]", "[nan, 1., 0.] rad"),
-        ([1.50, np.nan, 0] * u.rad, "[1.5rad nan 0rad]", "[1.5, nan, 0.] rad"),
+        ([np.nan, 1, 0] * u.rad, "[nan 1 rad 0 rad]", "[nan, 1., 0.] rad"),
```
Yeah, it is not looking good, but I cannot think of how to sort it out.

Telomelonia commented:
> Oops, indeed, this is more of a partial fix than I had realized! I think in principle there should be a space also for `latex`, etc. In #14231, I had [suggested](https://github.com/astropy/astropy/pull/14231/files#r1057999266) to use `super()` for the non-sexagesimal case, but a quick try caused more breakage than I had thought would be the case. Also `Quantity.to_string()` does not support `unicode` (see new issue #14386) and arguably does `latex` wrong in typesetting the degree symbol with a space in between it and the value (new issue #14385).
> 
> One option here may be to just keep it simple and focus on the generic format only -- i.e., only update this PR to also allow `format='generic'`.

I think one way is to include every format possible where space is required, since few units don't require space like `32'48''`  so there is too much if and else ... another possible way is to take the output from the function and check whether space should be given or not and if space is required then add 1 space else pass
The another task is to insert the space in the Latex part

mhvk commented:
My sense is still that apart from sexagesimal, this should all be handled by `Quantity.to_string()`. It may not be that tricky if we require that the `latex` and `unicode` entries for unit formats include whether a space is needed -- may need to edit things like Å and Ω -- or simply special-case the few that do not.

My sense, though, is not to let that stand in the way of doing the smaller change here, which solves things mostly.

mhvk commented:
@Telomelonia - I squashed your commits and added an extra one that also adds a space for other formats. I think this way it is ready to go in!

@eerovaher - would you mind having another look? I think your comment is now addressed.

mhvk commented:
Hopefully now got the `visualization.wcsaxis` bugs - should have tested locally with `matplotlib` installed...

pllim commented:
Are the "TODO"s in original post above still relevant?

mhvk commented:
@pllim - I think they're basically done with my additional commit, so I ticked them.

pllim commented:
Thanks, all!

Telomelonia commented:
> @Telomelonia - I squashed your commits and added an extra one that also adds a space for other formats. I think this way it is ready to go in!
> 
Thank you all for assisting me
